serverless-plugin-canary-deployments.js

```diff
@@ -59,7 +58,13 @@ class ServerlessCanaryDeployments {
     const functionAlias = this.getResourceLogicalName(aliasTpl);
     const lambdaPermissions = this.buildPermissionsForAlias({ functionName, functionAlias });
     const eventsWithAlias = this.buildEventsForAlias({ functionName, functionAlias });
-    return [deploymentGrTpl, aliasTpl, ...lambdaPermissions, ...eventsWithAlias];
+
+    const codeDeployRole = {};
+    if(typeof deploymentSettings.codeDeployRole === "undefined") {
+      Object.assign(codeDeployRole, this.buildCodeDeployRole());
+    }
+
```

Comment:
By shifting the role creation to the `.buildFunctionResources()` function, you'd be creating as many roles as functions with `deploymentSettings`, while we only need one. The `codeDeployRole` configuration should be global, not per function, so you could place it in the `custom` section of `serverless.yml`:

```yaml
custom:
  deploymentSettings:
    codeDeployRole: <...>
```

Actually, this section is already being [used for setting defaults](https://github.com/davidgf/serverless-plugin-canary-deployments/blob/master/serverless-plugin-canary-deployments.js#L227), although it is not documented (shame on me :sweat_smile: ). So, it would probably make sense to move the logic for generating the role back to the `.addCanaryDeploymentResources()` function, and check if there's a role configured in `.buildCodeDeployRole()`, returning an empty object it exists.